core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,304 @@
+
+  /**
+   * Interval at which heartbeat would be sent
+   */
+  private long interval;
+
+  /**
+   * Telemetry client instance used to send heartbeat.
+   */
+  private TelemetryClient telemetryClient;
+
+  /**
+   * ThreadPool used for adding properties to concurrent dictionary
+   */
+  private ExecutorService propertyUpdateService;
+
+  /**
+   * Threadpool used to send data heartbeat telemetry
+   */
+  private ScheduledExecutorService heartBeatSenderService;
+
+  /**
+   * Heartbeat enabled state
+   */
+  private volatile boolean isEnabled;
+
+  private final Object lock = new Object();
+
+  public HeartBeatProvider() {
+    this.interval = HeartBeatProviderInterface.DEFAULT_HEARTBEAT_INTERVAL;
+    this.heartbeatProperties = new ConcurrentHashMap<>();
+    this.isEnabled = true;
+    this.heartbeatsSent = new AtomicLong(0);
+    this.propertyUpdateService = Executors.newCachedThreadPool(ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+    this.heartBeatSenderService = Executors.newScheduledThreadPool(1, ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+  }
+
+  @Override
+  public String getInstrumentationKey() {
+    return this.telemetryClient.getContext().getInstrumentationKey();
+  }
+
+  @Override
+  public void setInstrumentationKey(String key) {
+    if (this.telemetryClient != null && key != null) {
+      this.telemetryClient.getContext().setInstrumentationKey(key);
+    }
+  }
+
+  @Override
+  public void initialize(TelemetryConfiguration configuration) {
+    if (isEnabled) {
+      if (this.telemetryClient == null) {
+        this.telemetryClient = new TelemetryClient(configuration);
+      }
+
+      //Submit task to set properties to dictionary using separate thread. we do not wait for the
+      //results to come out as some I/O bound properties may take time.
+      propertyUpdateService.submit(HeartbeatDefaultPayload.populateDefaultPayload(getExcludedHeartBeatProperties(),
+          getExcludedHeartBeatPropertyProviders(), this));
+
+      heartBeatSenderService.scheduleAtFixedRate(heartBeatPulse(), interval, interval, TimeUnit.SECONDS);
+    }
+  }
+
+  @Override
+  public boolean addHeartBeatProperty(String propertyName, String propertyValue,
+      boolean isHealthy) {
+
+    boolean isAdded= false;
+    if (!StringUtils.isEmpty(propertyName)) {
+      try {
+        if (!heartbeatProperties.containsKey(propertyName)) {
+             HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+             payload.setHealthy(isHealthy);
+             payload.setPayloadValue(propertyValue);
+             heartbeatProperties.put(propertyName, payload);
+             isAdded = true;
+             InternalLogger.INSTANCE.trace("added heartbeat property %s - %s", propertyName, propertyValue);
+        }
+        else {
+          throw new Exception("heartbeat property cannot be added twice. Please use setHeartBeatProperty instead to modify the value");
```

Comment:
Please use a more specific Exception subclass.

Reply:
This is nice to have but I don't think this should block the PR.